### ISSUE TITLE
SerialPort (win32) incompatible with gtk2reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4862]
|<img alt="detly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| detly reported|
|-|-|
|Trac ID|trac#4862|
|Type|defect|
|Created|2011-02-11 04:43:06Z|
|Branch|https://github.com/twisted/twisted/tree/win32-gtk-win32events-4862|

When an attempt is made to make a serial port connection under the gtk2reactor in Windows, the following traceback results:

```
File "C:\Python26\lib\site-packages\twisted\internet\_win32serialport.py", line 56, in __init__
    self.reactor.addEvent(self._overlappedRead.hEvent, self, 'serialReadEvent')
exceptions.AttributeError: 'PortableGtkReactor' object has no attribute 'addEvent'
```

The attached code take a port name as the first argument. It can be run under Linux, and doesn't do anything visible. Under Windows, you get the traceback.

Attachments:

* [tsgprob.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c32/c328e340dd966d895c1b00beecb7ccdf62c30eca/093512b3f4078184ca71630fb9519ad032877ca9.py) (428 bytes) - added by detly on 2011-02-11 04:43:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4862 4862
type__defect defect
reporter__detly detly
priority__normal normal
milestone__ 
branch__branches_win32_gtk_win32events_4862 branches/win32-gtk-win32events-4862
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1297399386000000 1297399386000000
changetime__1308007241000000 1308007241000000
version__None None
owner__exarkun exarkun
cc__detly cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|michaelnt commented|
|-|-|

I think this is a duplicate of [#3802](https://github.com/twisted/twisted/issues/3802), provide a better error message for reactors that don't support addEvent.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I think detly wants PortableGtkReactor to actually implement `IReactorWin32Events`.  It's certainly not too cool that you can't use gtk and serial ports together on Windows with Twisted right now.

**Comment by twisted-trac** [comment]
|<img alt="detly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|detly commented|
|-|-|

Just noticed [#3802](https://github.com/twisted/twisted/issues/3802) which is related, but exarkun is spot on in that I *can't* use the Win32 reactor, since it's a GTK app.

Back story: I was using a mess of threads and locking to get PyGTK and PySerial working together consistently under Windows and Linux. At the emphatic advice of #python, and much assistance from #twisted, I re-wrote my app to use Twisted instead. Since my dev environment is Linux, I didn't notice I'd sacrificed the ability to use the app on Windows — the main deployment environment in our workplace!

_tiny violin starts playing..._

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

So, the obvious fix is for `PortableGtkReactor` to implement `IReactorWin32Events`.  It would do that by calling `MsgWaitForMultipleObjects` in a separate thread, I guess.  Since `PortableGtkReactor` is just a thinly disguised `SelectReactor` it doesn't seem like it would be easy to get Windows event notification in the same thread.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [32008]) Branching to 'win32-gtk-win32events-4862'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [32009]) implement IReactorWin32Events for SelectReactor by giving it a Win32Reactor

refs [#4862](https://github.com/twisted/twisted/issues/4862)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

This seems to be largely complete in the branch.  However, there's some stickiness with `isInIOThread` which I haven't entirely sorted out yet.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/win32-gtk-win32events-4862).  Please review.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

  1. twisted/internet/selectreactor.py:18: 'IReactorWin32Events' imported but unused
  2. win32eventreactor._ThreadFDWrapper: Can we assume logPrefix is always thread-safe? (It *probably* is).
  3.  win32eventreactor._ThreadFDWrapper._execute: you should probably be using blockingCallFromThread, otherwise you're losing the remove-fd-if-action-handler-throws-exception behavior. That would mean connectionLost would be run in the thread, though, so you'd need to fix that too. Or you can recreate the exception handling in _execute, somehow.
  4. Add a test making sure that when the threaded win32 reactor shuts down it will *not* call connectionLost on anything (if there isn't one already, haven't gotten to tests yet.)
  5. The docstring for _ThreadFDWrapper should indicate the reactor it has is the non-win32 reactor, for clarity.

More comments later.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [32098]) Address the easy review points

refs [#4862](https://github.com/twisted/twisted/issues/4862)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [32099]) Avoid calling the logPrefix method in the wrong thread

refs [#4862](https://github.com/twisted/twisted/issues/4862)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [32100]) Dispatch the connectionLost callback to the main reactor thread

refs [#4862](https://github.com/twisted/twisted/issues/4862)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [32101]) Also pass on the handler's return value, in cause it is meant to cause the event source to disconnect.

refs [#4862](https://github.com/twisted/twisted/issues/4862)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Thanks very much.  Great points about the threaded interaction for `logPrefix` and `connectionLost`.  I think I addressed the issue, it wasn't so bad.  I just called `logPrefix` in the main thread and passed it along, rather than calling it each time in the secondary thread.  For exceptions raised from the application callback, `blockingCallFromThread` indeed gets the secondary reactor to disconnect the handler, and a suitable `callFromThread`-using `connectionLost` method on `_ThreadFDWrapper` ensures that application code can't tell there's an extra thread involved.

I didn't address point 4 yet.  There's something to be done there, I think, but I don't think we want to prevent `connectionLost` from being called - I think we want to ensure it is called at the right time?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst commented|
|-|-|

Re #4: Currently, AFAICT registering an event with win32eventreactor does *not* result in connectionLost on shutdown; that is only tied to read/write registration. connectionLost *will* be called from main reactor. So, there's nothing to do in actual code. However, we want to make sure this continues to be the case, thus my suggestion to add a test.

Changes look good. More issues, besides above:

  6. Remove my name as maintainer from modules you've edited :)
  7. Extra line break before _ThreadedWin32EventsMixin.
  8. If I set _registerAsIOThread to False in epoll, and then run "trial twisted.test.test_threadpool twisted.internet.test.test_threads", the test_isInIOThread incorrectly passes... because trial's reactor has already registered in the thread. Global state, it sucks.

More tomorrow, Jenny wants to go to sleep (sorry for the disjointed review.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

> Re #4: Currently, AFAICT registering an event with win32eventreactor does *not* result in connectionLost on shutdown; that is only tied to read/write registration. connectionLost *will* be called from main reactor. So, there's nothing to do in actual code. However, we want to make sure this continues to be the case, thus my suggestion to add a test.

Wow, terrible. :/  I guess I can add a test for that (r32108), it seems like stupid behavior though.

> Remove my name as maintainer from modules you've edited :)

r32107

> Extra line break before _ThreadedWin32EventsMixin.

Looks like 3 blank lines to me, that's right according to the coding standard.

> if I set _registerAsIOThread to False in epoll, and then run "trial twisted.test.test_threadpool twisted.internet.test.test_threads" ... Global state, it sucks.

Bleh, indeed.  I'm inclined to deprecate `isInIOThread` and, say, add a new reactor method.  Separate ticket, though.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_19"></a>|@itamarst set owner to @exarkun|
|-|-|

Re r32108, it's stupid behavior, yes, but mostly I was getting at we needed to make sure connectionLost wasn't called multiple times or in wrong thread.

Separate ticket for #8, and making it reactor method, seems like the way to go.

One remaining worry: if you forget to `implement(Win32EventsReactor)` in the right place, the relevant tests won't run and no one will notice that anything broke. Perhaps I'm worrying too much, though, so probably not necessary to do anything.

So unless you care about that: looks good, ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_20"></a>|@itamarst commented|
|-|-|

Oops, and review is done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

> Separate ticket for #8, and making it reactor method, seems like the way to go.

Filed [#5153](https://github.com/twisted/twisted/issues/5153), [#5154](https://github.com/twisted/twisted/issues/5154), and [#5155](https://github.com/twisted/twisted/issues/5155).

> One remaining worry: if you forget to implement(Win32EventsReactor) in the right place, the relevant tests won't run and no one will notice that anything broke.

True.  Something like this happened to cfreactor recently (and has not yet been fixed, I think).  This is a problem with the skip features of the `ReactorBuilder`-style tests though.  And I don't really know how to describe the problem (at least not in more detail than "sometimes stuff gets skipped that shouldn't get skipped") nor how we might fix it.  It's not a new problem with this branch, at least, so I'm happy merging this without addressing that problem.  I did manually verify that the all the tests that I want to be running are running before I submitted this for review, at least. :)

> So unless you care about that: looks good, ready to merge.

Yay, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set status to `closed`|
|-|-|

(In [32109]) Merge win32-gtk-win32events-4862 

Author: exarkun
Reviewer: itamarst
Fixes: [#4862](https://github.com/twisted/twisted/issues/4862)

Add `IReactorWin32Events` to the select, iocp, and Gtk reactors on Windows
(if the third-party dependencies are available) by having each of those
reactors run a `Win32Reactor` in a secondary thread and delegating event-
related methods to that reactor.

Significantly, this means all reactors on Windows now support serial ports.
